src/Test/StateMachine/Internal/Parallel.hs

```diff
@@ -94,29 +94,42 @@ generateParallelProgram generator precondition transition model = do
 -- | Shrink a parallel program in a pre-condition and scope respecting
 --   way.
 shrinkParallelProgram
-  :: HFoldable act
+  :: forall act model err
+  .  HFoldable act
   => Eq (Untyped act)
   => Shrinker act
   -> Precondition model act
   -> Transition' model act err
   -> model Symbolic
   -> (ParallelProgram act -> [ParallelProgram act])
 shrinkParallelProgram shrinker precondition transition model (ParallelProgram prefix suffixes)
-  = filter (validParallelProgram precondition transition model)
-      [ ParallelProgram (Program prefix') (map Program suffixes')
-      | (prefix', suffixes') <- shrinkPair'
-                                  (shrinkList (liftShrinkInternal shrinker))
-                                  (shrinkList (shrinkList (liftShrinkInternal shrinker)))
-                                  (unProgram prefix, map unProgram suffixes)
-      ] ++
-      case map unProgram suffixes of
-        []                   -> []
-        (suffix : suffixes') ->
-          [ ParallelProgram (prefix <> Program [prefix']) (Program suffix' : map Program suffixes')
-          | (prefix', suffix') <- pickOneReturnRest suffix
-          ]
-
+  = filter (validParallelProgram precondition transition model) $
+      [ ParallelProgram prefix' suffixes'
+      | (prefix', suffixes') <- shrinkPair' shrinkProgram' shrinkSuffixes (prefix, suffixes)
+      ]
+      ++
+      shrinkMoveSuffixToPrefix
   where
+  shrinkProgram' :: Program act -> [Program act]
+  shrinkProgram'
+    = map Program
+    . shrinkList (liftShrinkInternal shrinker)
+    . unProgram
+
+  shrinkSuffixes :: [(Program act, Program act)] -> [[(Program act, Program act)]]
+  shrinkSuffixes = shrinkList (shrinkPair shrinkProgram')
+
+  shrinkMoveSuffixToPrefix :: [ParallelProgram act]
+  shrinkMoveSuffixToPrefix = case suffixes of
+    []                   -> []
+    (suffix : suffixes') ->
+      [ ParallelProgram (prefix <> Program [prefix'])
+                        (bimap Program Program
+                              (splitAt (length suffix' `div` 2) suffix') : suffixes')
```

Comment:
Hmm this seems wrong to me, can't this potentially move things from one parallel part to the other?